### ISSUE TITLE
Built in string templating

### ISSUE THREAD

**Issue by conradludgate** [wjh/beaver#1]
Include expressions directly within strings and compile them to simple additions of strings.

For example

```bv
var name = "Conrad"
var age = 18
var foo = "Hello, My name is {name}. In 5 years I will be {age + 5} years old!"
```

Will become

```go
var name = "Conrad"
var age = 18
var foo = "Hello, My name is " + name + ". In 5 years I will be " + strconv.Itoa(age + 5) + " years old!"
```

The values within the brackets should be parsed as expressions.

Formatting of non-string values should be evaluated the same way were it to be parsed into fmt.Sprintf where all template values are "%v".  Therefore the following two statements should be equivalent.

```bv
var name = "Conrad"
var age = 18

var foo = "Hello, My name is {name}. In 5 years I will be {age + 5} years old!"
var bar = fmt.Sprintf("Hello, My name is %v. In 5 years I will be %v years old!", name, age + 5)
```

Processing the format during compile time would make the output much faster (just converting values to strings and concatenating them) than fmt.Sprintf, while also improving on the readability and size of the code.

Obviously, we would use '\\' to escape the open and close brackets.

**Comment by wjh** [comment]
Sounds good to me, we'll need to wait until I have a semantic pass written up however!

**Comment by conradludgate** [comment]
Just performed a benchmark, maybe not the most efficient but each method is used more or less how it's used in the wild. The results are as follows

https://gist.github.com/conradludgate/077ae4c150a980104d54c14c797cc8fb

```
BenchmarkConcat-4               20000000                94.7 ns/op
BenchmarkSprintf-4              10000000               233 ns/op
BenchmarkTextTemplate-4          2000000               725 ns/op
```

**Comment by wjh** [comment]
I guess this is a bit of a micro optimization, but on the other hand it'll add up.

If we go for what you proposed, default to concatenation unless the type is not a simple string. It'd be interesting to look into more advanced string templating like text/template provides, but by default it is a bit OP for simple string concatenation.

**Comment by conradludgate** [comment]
Yeah, Maybe with your proposed metatemplate programming, we could write a templater for the compiler to run instead. That way not all strings need to be processed, only the ones specified by the user. They could use simple fmt for debuging as it's faster to build but then they could use the MTP templates for the final release with the improved speeds.
